lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java

```diff
@@ -1221,6 +1228,110 @@ private Location createListingLocation(String path)
         return createLocation(path);
     }
 
+    @Test
+    void testFileDoesNotExistUntilClosed()
+            throws Exception
+    {
+        Location location = getRootLocation().appendPath("testFile2");
+        getFileSystem().deleteFile(location);
+        try (OutputStream out = getFileSystem().newOutputFile(location).create()) {
+            assertFalse(fileExistsInListing(location));
+            assertFalse(fileExists(location));
+            out.write("test".getBytes(UTF_8));
+            assertFalse(fileExistsInListing(location));
+            assertFalse(fileExists(location));
+        }
+        assertTrue(fileExistsInListing(location));
+        assertTrue(fileExists(location));
+        getFileSystem().deleteFile(location);
+    }
+
+    @Test
+    public void testLargeFileDoesNotExistUntilClosed()
+            throws IOException
+    {
+        Location location = getRootLocation().appendPath("testFile3");
+        getFileSystem().deleteFile(location);
+        byte[] bytes = new byte[8192];
+        for (int i = 0; i < bytes.length / 4; i++) {
+            //noinspection NumericCastThatLosesPrecision
+            bytes[i] = (byte) i;
+        }
+        try (OutputStream outputStream = getFileSystem().newOutputFile(location).create()) {
+            // write a 32 mb file
+            for (int i = 0; i < divide(32 * 1024 * 1024, bytes.length, UNNECESSARY); i++) {
```

Comment:
How about we do
```java
assertThat(MEGABYTE % bytes.length).isEqualTo(0);
// write a 33 MB file
int target = 33 * MEGABYTE;
int count = 0;
while (count < target) {
    outputStream.write(bytes);
    count += bytes.length;
    if (count % MEGABYTE == 0) {
        assertFalse(fileExistsInListing(location));
        assertFalse(fileExists(location));
    }
}

Reply:
Nice! They both take about 10-11seconds now.

Reply:
supercool. do we need to check every megabyte (33 times)?
would we lose anything if we asserted only before final flush/close?
ie do we believe it is possible for a file to first show up, then disappear, and then show up again while we're writing?